driver/getitem_driver.hpp

```diff
@@ -0,0 +1,594 @@
+#include "tensor_driver.hpp"
+#include "timer.hpp"
+#include "random.hpp"
+#include <algorithm>
+#include <cfloat>
+#include <cstdlib>
+#include <memory>
+#include <miopen/miopen.h>
+#include <miopen/tensor.hpp>
+#include <numeric>
+#include "tensor_view.h"
+#include <vector>
+#include <../test/tensor_holder.hpp>
+#include <../test/verify.hpp>
+
+tensor_view_5d_t get_inner_expanded_tv(const miopen::TensorDescriptor Desc)
+{
+    auto dims    = Desc.GetLengths();
+    auto strides = Desc.GetStrides();
+
+    tensor_view_5d_t tv_5d;
+    for(size_t i = 0; i < strides.size(); ++i)
+    {
+        tv_5d.stride[i] = strides[i];
+        tv_5d.size[i]   = dims[i];
+    }
+    auto rest = strides.size();
+    for(size_t j = rest; j < 5; ++j)
+    {
+        tv_5d.stride[j] = (rest == 0 ? 1 : strides[rest - 1]);
+        tv_5d.size[j]   = 1;
+    }
+    return tv_5d;
+}
+
+void slice_tv(tensor_view_5d_t& tv_5d, int32_t sliceCount, const int32_t* slices)
+{
+    for(int32_t i = 0; i < sliceCount; i++)
+    {
+        int32_t dim   = slices[4 * i + 0];
+        int32_t start = slices[4 * i + 1];
+        int32_t end   = slices[4 * i + 2];
+        int32_t step  = slices[4 * i + 3];
+
+        if(end > static_cast<int32_t>(tv_5d.size[dim]))
+            end = tv_5d.size[dim];
+
+        auto len = end - start;
+
+        tv_5d.size[dim] = (len + step - 1) / step;
+        tv_5d.stride[dim] *= step;
+    }
+}
+
+template <typename Tgpu, typename Tcheck>
+int32_t mloGetitemBackwardRunHost(miopenTensorDescriptor_t dyDesc,
+                                  int32_t indexCount,
+                                  miopenTensorDescriptor_t* indexDescs,
+                                  miopenTensorDescriptor_t dxDesc,
+                                  miopenTensorDescriptor_t errorDesc,
+                                  Tgpu* dy,
+                                  int32_t** indexs,
+                                  Tcheck* dxhost,
+                                  int32_t* errorhost,
+                                  int32_t dimCount,
+                                  int32_t* dims,
+                                  int32_t sliceCount,
+                                  int32_t* slices,
+                                  int32_t offset)
+{
+    auto dy_dims  = miopen::deref(dyDesc).GetLengths();
+    auto dy_numel = std::accumulate(dy_dims.begin(), dy_dims.end(), 1L, std::multiplies<int64_t>());
+    auto dx_dims  = miopen::deref(dxDesc).GetLengths();
+    auto index_dims = miopen::deref(indexDescs[0]).GetLengths();
+    auto index_numel =
+        std::accumulate(index_dims.begin(), index_dims.end(), 1L, std::multiplies<int64_t>());
+    auto element_index = std::vector<int32_t>(indexCount * index_numel + indexCount);
+
+    std::vector<int32_t> output_dims;
+    for(int32_t i = 0; i < dimCount; i++)
+    {
+        output_dims.push_back(dx_dims[dims[i]]);
+    }
+
+    auto dim_info_offset = indexCount > 0 ? indexCount * index_dims[0] : 0;
```

Comment:
signed/unsigned inconsistency: this line suggests that `indexCount` can be < 0 (otherwise, the >0 check is not necessary). However, the loop on line 125 will run forever if this is the case. We could either be more strict about it by checking `indexCount >= 0` and throwing an exception, or less strict and just add an assert to communicate our intention to the user. Personally, I think an assert is fine here:
```suggestion
    assert(indexCount >= 0);  // place this line at the top of the function
    auto dim_info_offset = indexCount * index_dims[0];
```

Reply:
Added a check when receiving driver input.

Reply:
@seungmanhan @randyspauldingamd I recommend using unsigned type for `dim_info_offset`, otherwise this kind of uncertainty (variable may be < 0) would propagate to the code that uses it.

Reply:
fixed it.